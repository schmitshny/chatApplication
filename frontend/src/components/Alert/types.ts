export type AlertType = 'success' | 'error' | 'info';
export interface Alert {
  id: string;
  message: string;
  type: AlertType;
}
