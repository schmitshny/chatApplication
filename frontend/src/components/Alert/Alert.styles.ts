import styled, { keyframes } from 'styled-components';
import { AlertType } from './types';

export const getAlertColor = (type: AlertType) => {
  switch (type) {
    case 'success':
      return '#4caf50';
    case 'error':
      return '#f44336';
    case 'info':
      return '#2196f3';
  }
};

export const StyledAlertContainer = styled.div`
  position: fixed;
  top: 20px;
  right: 20px;
  z-index: 1000;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  gap: 10px;
`;

export const AlertBox = styled.div`
  position: relative;
  background-color: ${({ theme }) => theme.colors.contrastBackground};
  color: ${({ theme }) => theme.colors.contrastText};
  padding: 10px 15px 15px;
  border-radius: 4px;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.2);
  display: flex;
  justify-content: space-between;
  align-items: center;
  min-width: 300px;
  box-sizing: border-box;
  animation: slideIn 0.5s ease;
  overflow: hidden;
`;

export const AlertMessage = styled.span`
  flex-grow: 1;
  padding: 0 20px 0 10px;
`;

export const CloseIcon = styled.div`
  color: #666;
  position: absolute;
  top: 4px;
  right: 4px;

  &:hover {
    color: #000;
  }
`;

const shrink = keyframes`
  from {
    width: 100%;
  }
  to {
    width: 0%;
  }
`;

export const ProgressBar = styled.div<{ type: AlertType }>`
  height: 5px;
  background-color: ${({ type }) => getAlertColor(type)};
  animation: ${shrink} 3s linear forwards;
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
`;
