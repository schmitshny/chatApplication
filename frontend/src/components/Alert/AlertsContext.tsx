import { createContext } from 'react';
import { Alert, AlertType } from './types';

interface AlertContextType {
  alerts: Alert[];
  addAlert: (message: string, type: AlertType) => void;
  removeAlert: (id: string) => void;
}

export const AlertContext = createContext<AlertContextType | null>(null);
