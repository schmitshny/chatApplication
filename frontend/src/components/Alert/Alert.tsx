import { Icon } from '..';
import { AlertBox, AlertMessage, CloseIcon, ProgressBar, getAlertColor } from './Alert.styles';
import { AlertType } from './types';

interface AlertProps {
  message: string;
  onClose: () => void;
  type: AlertType;
}

export const Alert = ({ message, onClose, type }: AlertProps) => {
  return (
    <AlertBox>
      <Icon name="alert" customColor={getAlertColor(type)} />
      <AlertMessage>{message}</AlertMessage>
      <CloseIcon>
        <Icon onClick={onClose} name="delete" data-testid="close-alert-icon" />
      </CloseIcon>
      <ProgressBar type={type} />
    </AlertBox>
  );
};
