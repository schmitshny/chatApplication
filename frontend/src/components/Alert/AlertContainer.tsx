import { useAlerts } from './useAlerts';
import { Alert } from './Alert';
import { StyledAlertContainer } from './Alert.styles';

export const AlertContainer = () => {
  const { alerts, removeAlert } = useAlerts();

  return (
    alerts.length > 0 && (
      <StyledAlertContainer>
        {alerts.map(({ id, message, type }) => (
          <Alert key={id} message={message} onClose={() => removeAlert(id)} type={type} />
        ))}
      </StyledAlertContainer>
    )
  );
};
