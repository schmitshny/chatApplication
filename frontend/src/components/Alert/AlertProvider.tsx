import { useState, useEffect, ReactNode } from 'react';
import { Alert, AlertType } from './types';
import { AlertContext } from './AlertsContext';

interface AlertProviderProps {
  children: ReactNode;
}

export const AlertProvider = ({ children }: AlertProviderProps) => {
  const [alerts, setAlerts] = useState<Alert[]>([]);

  const addAlert = (message: string, type: AlertType) => {
    const id = Math.random().toString(36).substring(2, 9);
    setAlerts((prevAlerts) => [...prevAlerts, { id, message, type }]);
  };

  const removeAlert = (id: string) => {
    setAlerts((prevAlerts) => prevAlerts.filter((alert) => alert.id !== id));
  };

  useEffect(() => {
    if (alerts.length > 0) {
      const timer = setTimeout(() => {
        removeAlert(alerts[0].id);
      }, 3000);

      return () => clearTimeout(timer);
    }
  }, [alerts]);

  return <AlertContext.Provider value={{ alerts, addAlert, removeAlert }}>{children}</AlertContext.Provider>;
};
