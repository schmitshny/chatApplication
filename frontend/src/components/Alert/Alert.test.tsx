import { describe, expect, test, vi } from 'vitest';
import { screen, fireEvent } from '@testing-library/react';
import { render } from '../../tests/test-utils';
import { Alert } from './Alert';

const mockOnClose = vi.fn();

describe('Alert Component', () => {
  test('renders with the correct message', () => {
    const testMessage = 'Test Alert Message';
    const { container } = render(<Alert message={testMessage} onClose={mockOnClose} type="error" />);

    expect(screen.getByText(testMessage)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('calls onClose when the close icon is clicked', () => {
    render(<Alert message="Test Alert Message" onClose={mockOnClose} type="error" />);

    const closeIcon = screen.getByTestId('close-alert-icon');
    fireEvent.click(closeIcon);

    expect(mockOnClose).toHaveBeenCalled();
  });
});
