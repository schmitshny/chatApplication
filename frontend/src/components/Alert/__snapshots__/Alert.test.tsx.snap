// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert Component > renders with the correct message 1`] = `
<div>
  <div
    class="sc-llJbZy yVbvn"
  >
    <div
      class="sc-jSMdHm iLFJpR"
    >
      <svg
        fill="red"
        height="20px"
        id="Capa_1"
        stroke="red"
        version="1.1"
        viewBox="0 0 310.806 310.806"
        width="20px"
        xml:space="preserve"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <g
          id="SVGRepo_bgCarrier"
          stroke-width="0"
        />
        <g
          id="SVGRepo_tracerCarrier"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
        <g
          id="SVGRepo_iconCarrier"
        >
          <path
            d="M305.095,229.104L186.055,42.579c-6.713-10.52-18.172-16.801-30.652-16.801c-12.481,0-23.94,6.281-30.651,16.801 L5.711,229.103c-7.145,11.197-7.619,25.39-1.233,37.042c6.386,11.647,18.604,18.883,31.886,18.883h238.079 c13.282,0,25.5-7.235,31.888-18.886C312.714,254.493,312.24,240.301,305.095,229.104z M155.403,253.631 c-10.947,0-19.82-8.874-19.82-19.82c0-10.947,8.874-19.821,19.82-19.821c10.947,0,19.82,8.874,19.82,19.821 C175.223,244.757,166.349,253.631,155.403,253.631z M182.875,115.9l-9.762,65.727c-1.437,9.675-10.445,16.353-20.119,14.916 c-7.816-1.161-13.676-7.289-14.881-14.692l-10.601-65.597c-2.468-15.273,7.912-29.655,23.185-32.123 c15.273-2.468,29.655,7.912,32.123,23.185C183.284,110.192,183.268,113.161,182.875,115.9z"
          />
           
        </g>
      </svg>
    </div>
    <span
      class="sc-iIPlFl kHauvF"
    >
      Test Alert Message
    </span>
    <div
      class="sc-gicDKM eDUHUO"
    >
      <div
        class="sc-jSMdHm kMbLdu"
        data-testid="close-alert-icon"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 5C15.9 5 19 8.1 19 12C19 15.9 15.9 19 12 19C8.1 19 5 15.9 5 12C5 8.1 8.1 5 12 5ZM12 4C7.6 4 4 7.6 4 12C4 16.4 7.6 20 12 20C16.4 20 20 16.4 20 12C20 7.6 16.4 4 12 4Z"
            fill="#c71111"
            fill-opacity="1"
          />
          <path
            d="M16.2 14.8L13.4 12L16.2 9.20005L14.8 7.80005L12 10.6L9.19999 7.80005L7.79999 9.20005L10.6 12L7.79999 14.8L9.19999 16.2L12 13.4L14.8 16.2L16.2 14.8Z"
            fill="#eb5a5a"
          />
        </svg>
      </div>
    </div>
    <div
      class="sc-ezWNbD cgQkIn"
      type="error"
    />
  </div>
</div>
`;
